### ISSUE TITLE
Decouple metrics submission from RUST_LOG

### ISSUE THREAD

**Issue by t-nelson** [solana-labs/solana#15215]
#### Problem

It is occasionally desired to reduce text log size, but continue submitting full metrics.  With metrics level taken from `RUST_LOG`, this is not possible

#### Proposed Solution

Look into the feasibility of decoupling metrics from `RUST_LOG`

**Comment by meyerbro** [comment]
Hey @t-nelson any chance for this to be implemented before August/2022? This will impact a lot of people now that metrics are mandatory. Thanks in advance!

**Comment by Erfankam** [comment]
@t-nelson I would like to tackle on this issue. Would you mind to assigning this issue to me?

**Comment by t-nelson** [comment]
> @t-nelson I would like to tackle on this issue. Would you mind to assigning this issue to me?

it's yours!

**Comment by Erfankam** [comment]
> > @t-nelson I would like to tackle on this issue. Would you mind to assigning this issue to me?
> 
> it's yours!

I m preparing my environment.

**Comment by Erfankam** [comment]
I worked on this issue for a while by reading source codes and config files in samples and at end of the day, by trying to change some metrics command s function signature.
I think decoupling metrics logs from `RUST_LOGS` is possible through out these solutions:

1. Using **per module logging** mechanism: As `metrics` is a module, we can decouple metrics logs from general logs (containing other modules and core) by specifying metrics log level in RUST_LOG environment specifically. Using this method, we have to do nothing except using different RUST_LOG environment based on user needs. For example:
`export RUST_LOG=info,solana-metrics=warn`
In this way, we give the freedom to users to consider compound or separated log level configurations.
2. Using **new configuration**: We can define new configuration key in `core` or `logs `module to be read from config files and inject specific log level into metrics module.
In this way, we should **_define another logger_** other than the general one to consider specific log level.
As a user, I prefer to not using or knowing another configuration. So, **I think solution 1 is better**.
@t-nelson Am I right? If not, please correct me?

**Comment by t-nelson** [comment]
i was thinking along the lines of an analog to `RUST_LOG`, but for metrics. probably falling back to `RUST_LOG` if not specified.

so i could have something like 
```
export RUST_LOG=solana=warn,...
export SOLANA_METRICS=solana=info,...
```

is that your proposal 1?

**Comment by Erfankam** [comment]
Yeah. exactly. But my **first proposal** is to just using RUST_LOG like following:
`export RUST_LOG=solana=warn,solana-metrics=debug`

I think this given requirement (Decouple metrics submission from RUST_LOG) **_is already provided_** by using suitable SOLANA_LOG environment variable value and **without need to define any other configuration or environment** variable like witch you mentioned. Does not it?

**Comment by t-nelson** [comment]
what is `SOLANA_LOG`?  it hasn't been introduced by that name here, nor does it exist in the codebase today.

i honestly have no idea what your proposal is, practically speaking

**Comment by Erfankam** [comment]
> what is `SOLANA_LOG`? it hasn't been introduced by that name here, nor does it exist in the codebase today.
> 
> i honestly have no idea what your proposal is, practically speaking

Sorry. It was my fault and It was just a misspelling. I revised that. Would you please to read it again?

**Comment by steviez** [comment]
> I think this given requirement (Decouple metrics submission from RUST_LOG) is already provided by using suitable RUST_LOG environment variable value and without need to define any other configuration or environment variable like witch you mentioned. Does not it?

Did you test your theory; I don't think it works like you think it does. The following snippet is where all the datapoints go through:
https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/datapoint.rs#L153-L165

The log level is checked at this level, so setting `RUST_LOG=solana=info,solana-metrcis=debug` will disable the call to `submit()` for any `datapoint_info!`s. If you follow the call-stack down, items passed to `submit()` are eventually processed here:

https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/metrics.rs#L261-L264

The first line logs them (prints to stderr or file), the second line pushes them in a `Vec` to be handled later.

**Comment by Erfankam** [comment]
I will check it out and inform you soon.

**Comment by steviez** [comment]
I was already running some nodes, so I set `RUST_LOG` one node to confirm understanding:
```
RUST_LOG=solana=info,solana_metrcis=warn
```
As expected, the node's logs didn't have any of the following log lines; just several `WARN` logs (as expected).
```
[<time> INFO  solana_metrics::metrics] datapoint: <datapoint> <fields>
```
But, checking the metrics explorer, my node was indeed submitting metrics this whole time.

Going back to code, the `datapoint_<level>!` invocations are macros, **so those are getting expanded in their calling locations**. If `datapoint_info!` is called from `solana_core`, the log level of `solana_core` is what gates whether a datapoint makes it past the below check, not `solana_metrics` level.

https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/datapoint.rs#L153-L165

So, a `datapoint_info` (which has level INFO) makes it through the above `log_enabled!` check, and we end up hitting the other snippet I had linked:

https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/metrics.rs#L261-L264

Items are pushed into the `Vec` regardless, so if we make it here, level is now irrelevant. For the call to `log!()`, the level will get [checked again](https://docs.rs/log/0.4.17/src/log/macros.rs.html#31), except, we are now in a `solana_metrics` function.

So, in my above example, a `datapoint_info!()` call has level `info`, and since `solana_metrics` was configured with `warn`, the statement is not emitted to my log file.

So, in a flip from my previous comment, I think setting `RUST_LOG` to below is probably sufficient for the majority of cases:
```
RUST_LOG=solana=info,solana_metrcis=warn
```
We do have some metrics that are at lower levels, so hypothetically, there could be a want for enabling metrics of lower level (ie debug) without adjusting the log level of the entire crate. This seems pretty niche tho, and anyone wanting to do this could just rebuild from source and adjust the level themself. If we want the entire cluster to begin reporting a metric, we're better off pushing a release with the level adjusted as opposed to asking all operators to adjust an env variable.

TLDR of this long post - I think we can likely close this issue with no code changes. Maybe a documentation adjustment in the docs.

**Comment by Erfankam** [comment]
> > I think this given requirement (Decouple metrics submission from RUST_LOG) is already provided by using suitable RUST_LOG environment variable value and without need to define any other configuration or environment variable like witch you mentioned. Does not it?
> 
> Did you test your theory; I don't think it works like you think it does. The following snippet is where all the datapoints go through:
> 
> https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/datapoint.rs#L153-L165
> 
> The log level is checked at this level, so setting `RUST_LOG=solana=info,solana-metrcis=warn` will disable the call to `submit()` for any `datapoint_info!`s. If you follow the call-stack down, items passed to `submit()` are eventually processed here:
> 
> https://github.com/solana-labs/solana/blob/d4ce59eee7e2537f02c827b6918de9af559b06bd/metrics/src/metrics.rs#L261-L264
> 
> The first line logs them (prints to stderr or file), the second line pushes them in a `Vec` to be handled later.
> 
> Edit: Added the code links to the problem description above to help guide readers into looking at the correct spot.

@steviez 
I checked it in my environment. You know, because these macros are reused inside other modules than metrics, so the host module log level will take effect and metrics log will just affect on submit calls.
So, We can still use this mechanism (Using RUST_LOG) for decoupling metrics logs and other module log levels.

**Comment by Erfankam** [comment]
> RUST_LOG=solana=info,

As you see, we agree that decoupling metrics is already works by using RUST_LOG environment variable. I think although decoupling metrics and logging is a good idea, but adding another configuration not worth it.
